File: _REPO/APRESS/distributed-.net-programming-in-vb-.net/Chapter08/ObjectPooling/SimpleCarLibrary/CarService.vb
'All the important enterprise service types are contained here.
Imports System.EnterpriseServices
Imports System.Runtime.InteropServices 'For GUID attribute
Imports System.Text 'For StringBuilder

<Guid("5A1BB09D-8D4B-481a-A441-DC6D8D48F396")> _
Public Interface IGetInfo
   Function GetInfo() As String
   Sub AnotherMethod()
End Interface

<EventTrackingEnabled(), _
 JustInTimeActivation(), _
 ClassInterface(ClassInterfaceType.None), _
 ObjectPooling(5, 10)> _
Public Class CarService
   Inherits ServicedComponent
   Implements IGetInfo
      
   Public Sub New
      Dim logMsg As String = string.Format("Created Object: {0}", _
         Me.GetHashCode())
      EventLog.WriteEntry("CarService.ctor", logMsg)
   End Sub   

   Protected Overrides Sub Activate()
      Dim logMsg As String = String.Format("Activated Object: {0}", _
         Me.GetHashCode())
      EventLog.WriteEntry("CarService.Activate", logMsg)
   End Sub

   Protected Overrides Function CanBePooled() As Boolean
      Dim logMsg As String = string.Format("Object {0} not pooled", _
         Me.GetHashCode())
      EventLog.WriteEntry("CarService.CanBePooled", logMsg)
      Return True
   End Function

   Protected Overrides Sub Deactivate()
      Dim logMsg As String = String.Format("Deactivated Object: {0}", _
         Me.GetHashCode())
      EventLog.WriteEntry("CarService.Deactivate", logMsg)
   End Sub
   
   Protected Overrides Sub Finalize()
      Dim logMsg As String = String.Format("Object {0} Finalized", _
         Me.GetHashCode())
      EventLog.WriteEntry("CarService.Finalize", logMsg)
      MyBase.Finalize()
   End Sub
   
   Public Function GetInfo() As String Implements IGetInfo.GetInfo
      Dim ctxInfo As New StringBuilder()

      'Use ContextUtil to fetch context information
      ctxInfo.AppendFormat("Context ID:  {0}{1}", _
         ContextUtil.ContextId, vbNewline)
      ctxInfo.AppendFormat("Activity ID: {0}{1}", _
         ContextUtil.ActivityId, vbCrLf)

      'If in transaction, get transaction ID
      Dim txId As String = "No Tx"
      If ContextUtil.IsInTransaction Then
         txId = ContextUtil.TransactionId.ToString()
      End If

      ctxInfo.AppendFormat("Transaction ID:    {0}{1}", txId, vbCrLf)
      ctxInfo.AppendFormat("Security Enabled?: {0}{1}", _
         ContextUtil.IsSecurityEnabled, vbCrLf)
         
      ContextUtil.SetComplete()
         
      Return ctxInfo.ToString()
   End Function
   
   Public Sub AnotherMethod() Implements IGetInfo.AnotherMethod
      ContextUtil.SetComplete()
   End Sub   
End Class

File: _REPO/APRESS/distributed-.net-programming-in-csharp/Chapter08/ObjectPooling/SimpleCarLibrary/CarService.cs
using System;
using System.Diagnostics;              // event Logging
using System.EnterpriseServices;       // ServicedComponent
using System.Runtime.InteropServices;  // Class Interface
using System.Text;                     // StringBuilder
using System.Threading;

namespace SimpleCarLibrary
{
   [Guid("5A1BB09D-8D4B-481a-A441-DC6D8D48F396")]
   public interface IGetInfo
   {
      string GetInfo();
      void AnotherMethod();
   }

   [EventTrackingEnabled()]
   [JustInTimeActivation()]
   [ClassInterface(ClassInterfaceType.None)]
   [ObjectPooling(5, 10)]
   public class CarService : ServicedComponent, IGetInfo
   {
      public CarService()
      {
         string logMsg = string.Format("Created Object: {0}", this.GetHashCode());
         EventLog.WriteEntry("CarService.ctor", logMsg);
      }

      public string GetInfo()
      {
         StringBuilder ctxInfo = new StringBuilder();
         
         // Use ContextUtil to fetch context information
         ctxInfo.AppendFormat("Context ID:  {0}\n", ContextUtil.ContextId);
         ctxInfo.AppendFormat("Activity ID: {0}\n", ContextUtil.ActivityId);
        
         // If in transaction, get transaction ID
         string txId = "No Tx";
         if (ContextUtil.IsInTransaction)
            txId = ContextUtil.TransactionId.ToString();

         ctxInfo.AppendFormat("Transaction ID:    {0}\n", txId);
         ctxInfo.AppendFormat("Security Enabled?: {0}\n", 
            ContextUtil.IsSecurityEnabled);

         // Tell COM+ to deactivate the object upon return
         ContextUtil.SetComplete();

         return ctxInfo.ToString();
      }
      
      public void AnotherMethod()
      {
         ContextUtil.SetComplete();
      }

      protected override void Activate()
      {
         string logMsg = string.Format("Activated Object: {0}", 
            this.GetHashCode());
         EventLog.WriteEntry("CarService.Activate", logMsg);
      }
      protected override void Deactivate()
      {
         string logMsg = string.Format("Deactivated Object: {0}", 
            this.GetHashCode());
         EventLog.WriteEntry("CarService.Deactivate", logMsg);
      }
      protected override bool CanBePooled()
      {
         string logMsg = string.Format("Object {0} pooled", 
            this.GetHashCode());
         EventLog.WriteEntry("CarService.CanBePooled", logMsg);
         return true;
      }

      ~CarService()
      {
         string logMsg = string.Format("Object {0} Finalized", 
            this.GetHashCode());
         EventLog.WriteEntry("CarService.Finalize", logMsg);
      }
   }
}